erc20.go

```diff
@@ -0,0 +1,363 @@
+		TxSequenceID uint16
+	}
+)
+
+// NewERC20Plugin creates a new erc20 plugin
+func NewERC20Plugin(ERC20FeePoolAddress rivinetypes.UnlockHash, oneCoin rivinetypes.Currency, erc20TxValidator types.ERC20TransactionValidator) *Plugin {
+	p := &Plugin{}
+	rivinetypes.RegisterTransactionVersion(types.TransactionVersionERC20Conversion, types.ERC20ConvertTransactionController{})
+	rivinetypes.RegisterTransactionVersion(types.TransactionVersionERC20CoinCreation, types.ERC20CoinCreationTransactionController{
+		Registry:    p,
+		OneCoin:     oneCoin,
+		TxValidator: erc20TxValidator,
+	})
+	rivinetypes.RegisterTransactionVersion(types.TransactionVersionERC20AddressRegistration, types.ERC20AddressRegistrationTransactionController{
+		Registry:             p,
+		OneCoin:              oneCoin,
+		BridgeFeePoolAddress: ERC20FeePoolAddress,
+	})
+	return p
+}
+
+// InitPlugin initializes the Bucket for the first time
+func (p *Plugin) InitPlugin(metadata *persist.Metadata, bucket *bolt.Bucket, storage modules.PluginViewStorage, unregisterCallback modules.PluginUnregisterCallback) (persist.Metadata, error) {
+	p.storage = storage
+	p.unregisterCallback = unregisterCallback
+	if metadata == nil {
+		buckets := [][]byte{
+			bucketERC20ToTFTAddresses,
+			bucketTFTToERC20Addresses,
+			bucketERC20TransactionIDs,
+		}
+		var err error
+		for _, bckt := range buckets {
+			_, err = bucket.CreateBucketIfNotExists(bckt)
+			if err != nil {
+				return persist.Metadata{}, err
+			}
+		}
+
+		metadata = &persist.Metadata{
+			Version: pluginDBVersion,
+			Header:  pluginDBHeader,
+		}
+	} else if metadata.Version != pluginDBVersion {
+		return persist.Metadata{}, errors.New("There is only 1 version of this plugin, version mismatch")
+	}
+	return *metadata, nil
+}
+
+// ApplyBlock applies a block's erc20 transaction to their buckets.
+func (p *Plugin) ApplyBlock(block rivinetypes.Block, height rivinetypes.BlockHeight, bucket *persist.LazyBoltBucket) error {
+	if bucket == nil {
+		return errors.New("erc20 bucket does not exist")
+	}
+	tx, err := bucket.Tx()
+	if err != nil {
+		return errors.New("erc20 bucket does not exist")
+	}
+	for i := range block.Transactions {
+		rtx := &block.Transactions[i]
+		if rtx.Version == rivinetypes.TransactionVersionOne {
+			continue // ignore most common Tx
+		}
+		ctx := transactionContext{
+			BlockHeight:  height,
+			BlockTime:    block.Timestamp,
+			TxSequenceID: uint16(i),
+		}
+		// check the version and handle the ones we care about
+		switch rtx.Version {
+		case types.TransactionVersionERC20CoinCreation:
+			err = p.applyERC20CoinCreationTx(tx, ctx, rtx)
+		case types.TransactionVersionERC20AddressRegistration:
+			err = p.applyERC20AddressRegistrationTx(tx, ctx, rtx)
+		}
+	}
+	return nil
+}
+
+// RevertBlock reverts a block's erc20 transaction from their buckets
+func (p *Plugin) RevertBlock(block rivinetypes.Block, height rivinetypes.BlockHeight, bucket *persist.LazyBoltBucket) error {
+	if bucket == nil {
```

Comment:
should not be required though? Why would this function ever be called with a nil bucket? can that happen from the rivine side? You could also turn it into an interface of course so that the writer of a plugin doesn't get scared of the pointer.